Source/NameServer/AddressBook/AddressBook.cpp

```diff
@@ -0,0 +1,1367 @@
+    // Add the device to the correct Supervisor list.
+    // Supervisors may be loaded after devices, may need to build the map.
+    SupMap_t::iterator SSearch = TRec->SupMap.find(DRec->Supervisor);
+    if (SSearch == TRec->SupMap.end()) { //Supervisor is NOT in the map yet.
+        std::pair<SupMap_t::iterator,bool> LastVal;     // Have to do this because C++98 does not like initialiser lists...
+        LastVal = TRec->SupMap.insert(SupMap_t::value_type(DRec->Supervisor, RecordVect_t()));
+        LastVal.first->second.push_back(DRec);  // RecordVect_t({DRec}) in the insert would be neater and not need the pair.
+
+    } else {                             //Supervisor is already in the map.
+        SSearch->second.push_back(DRec);
+    }
+
+    return SUCCESS;
+}
+
+
+/*==============================================================================
+ * ValidateTask:    Check that the task has a valid vector of DeviceTypes.
+ *============================================================================*/
+unsigned AddressBook::ValidateTask(TaskData_t &Data)
+{
+    for(std::vector<DevTypeRecord_t>::const_iterator D=Data.DeviceTypes.begin();
+            D!=Data.DeviceTypes.end(); D++)
+    {
+        if (unsigned retVal = ValidateDeviceType(*D, Data.MessageTypes.size()))
+            return retVal;
+    }
+
+    return SUCCESS;
+}
+
+/*==============================================================================
+ * ValidateDeviceType:  Check that DeviceType Message indices are valid and not 
+ *                  out of range.
+ * Added and separated from ValidateTask 5 July 2019 ADR
+ *============================================================================*/
+unsigned AddressBook::ValidateDeviceType(const DevTypeRecord_t& DevTypRec, unsigned MaxIdx)
+{
+    for(std::vector<MsgIdx>::const_iterator I=DevTypRec.InMsgs.begin();
+        I!=DevTypRec.InMsgs.end(); I++)
+    {
+        if(*I >= MaxIdx)
+        {
+            return(ERR_INVALID_MESSAGE_TYPE);
+        }
+    }
+
+    // changed to iterate over OuMsgs (was InMsgs) 5 July 2019 ADR
+    for(std::vector<MsgIdx>::const_iterator O=DevTypRec.OuMsgs.begin();
+        O!=DevTypRec.OuMsgs.end(); O++)
+    {
+        if(*O >= MaxIdx)
+        {
+            return(ERR_INVALID_MESSAGE_TYPE);
+        }
+    }
+    return 0;
+}
+
+
+/*==============================================================================
+ * ValidateDevice:  Check that the device does not already exist in the name map
+ *                  and that indices in the Record are not out of range.
+ *============================================================================*/
+unsigned AddressBook::ValidateDevice(TaskRecord_t *TRec, Record_t &DevRec)
+{
+    // Check that the device does not already exist in the maps
+    AddrMap_t::const_iterator DASearch = TRec->AddrMap.find(DevRec.Address);
+    if (DASearch != TRec->AddrMap.end()) {
+        // DebugPrint("Device address %llu already exists\n", DevRec.Address);
```

Comment:
Could enable these off a preprocessor define if they regularly come in handy for debugging.